include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,366 @@
+        unsigned long long ctr_inc = z;
+
+        for (size_t i = 0; i < word_count; ++i) {
+            scalar_type prv_x = state_.X[i];
+            state_.X[i] = ((state_.X[i] + ctr_inc&in_mask) & in_mask) + carry;
+            carry = 0;
+
+            if(state_.X[i] < prv_x) { // overflow of the chunk
+                carry = 1;
+            }
+
+            ctr_inc = (ctr_inc&(~in_mask))>>(sizeof(z)*8-word_size);
+        }
+    }
+    
+    /* operator() specified for sycl_vec output */
+    template <unsigned int _N = 0>
+    ::std::enable_if_t<(_N > 0), result_type>
+    operator()() {
+        result_type tmp;
+        scalar_type curr_idx = state_.idx;
+
+        for (int elm_count = 0; elm_count < _N; elm_count++, curr_idx++) {
+            if(curr_idx  == word_count) { // empty buffer
+                generate();
+                increase_counter_internal();
+                curr_idx = 0;
+            }
+            tmp[elm_count] = state_.Y[curr_idx];
+        }
+        
+        state_.idx = ++curr_idx;
+
+        return tmp;
+    }
+    
+    /* operator() specified for a scalar output */
+    template <unsigned int _N = 0>
+    ::std::enable_if_t<(_N == 0), result_type>
+    operator()() {
+        result_type tmp;
+
+        scalar_type curr_idx = state_.idx;
+        if(curr_idx  == word_count) { // empty buffer
+            generate();
+            increase_counter_internal();
+            curr_idx = 0;
+        }
+        
+        // There are already generated numbers in the buffer
+        tmp = state_.Y[curr_idx];
+        state_.idx = ++curr_idx;
+
+        return tmp;
+    }
+
+public:
+    static constexpr ::std::array<scalar_type, array_size> multipliers =
+        detail::get_even_array_from_tuple<scalar_type>(::std::make_tuple(consts...),
+                                                    ::std::make_index_sequence<array_size>{});
+    static constexpr ::std::array<scalar_type, array_size> round_consts =
+        detail::get_odd_array_from_tuple<scalar_type>(::std::make_tuple(consts...),
+                                                    ::std::make_index_sequence<array_size>{});
+    static constexpr scalar_type min() { return 0; }
+    static constexpr scalar_type max() { return ::std::numeric_limits<scalar_type>::max(); }
+    static constexpr scalar_type default_seed = 20111115u;
+
+    /* Constructors and seeding functions */
+    philox_engine() : philox_engine(default_seed) {}
+    explicit philox_engine(scalar_type value) { seed(value); }
+    void seed(scalar_type value = default_seed) { seed_internal({ value & in_mask }); }
+
+    /* Set the state to arbitrary position */
+    void set_counter(const ::std::array<scalar_type, word_count>& counter) {
+        auto start = counter.begin();
+        auto end = counter.end();
+        for (size_t i = 0; i < word_count; i++) {
+            // all counters are set in everse order
+            state_.X[i] = (start == end) ? 0 : (*--end) & in_mask;
```

Comment:
I assume we don't have to check start == end here as counter is guaranteed to have word_count size.

Reply:
Hm, sounds reasonable, removed.